File: Project-source/forge-verifier/csharp/src/SoftFingerprint.cs
// Soft fingerprint — SHA-256 over (MAC | hostname | CPU). Matches the canonical
// string format used by the Go / Python / Java SDKs so a license bound on one
// host can be verified by any-language verifier on the same host.

using System.Net.NetworkInformation;
using System.Net.Sockets;
using System.Security.Cryptography;
using System.Text;

namespace YourCo.Forge.Verifier;

public static class SoftFingerprint
{
    public static string Compute()
    {
        var mac = Safe(PrimaryMac());
        var hostname = Safe(Environment.MachineName);
        var cpu = Safe(CpuDescription());

        var canonical = $"{mac}|{hostname}|{cpu}".ToLowerInvariant();
        var hash = SHA256.HashData(Encoding.UTF8.GetBytes(canonical));
        return Convert.ToHexString(hash).ToLowerInvariant();
    }

    private static string PrimaryMac()
    {
        foreach (var ni in NetworkInterface.GetAllNetworkInterfaces())
        {
            if (ni.OperationalStatus != OperationalStatus.Up) continue;
            if (ni.NetworkInterfaceType == NetworkInterfaceType.Loopback) continue;
            var name = ni.Name.ToLowerInvariant();
            if (name.StartsWith("docker") || name.StartsWith("veth")
                || name.StartsWith("vmnet") || name.StartsWith("br-")
                || name.StartsWith("utun")) continue;
            // Skip tunneling / virtual-ish types
            if (ni.NetworkInterfaceType is NetworkInterfaceType.Tunnel
                or NetworkInterfaceType.Unknown) continue;

            var bytes = ni.GetPhysicalAddress().GetAddressBytes();
            if (bytes.Length == 0) continue;
            return string.Join(":", bytes.Select(b => b.ToString("x2")));
        }
        return "";
    }

    private static string CpuDescription() =>
        $"{System.Runtime.InteropServices.RuntimeInformation.ProcessArchitecture}:" +
        $"{System.Runtime.InteropServices.RuntimeInformation.OSDescription}:" +
        $"{Environment.ProcessorCount}";

    private static string Safe(string? s) =>
        string.IsNullOrEmpty(s) ? "unknown" : s;
}

File: Project-source/forge-verifier/java/src/main/java/com/yourco/forge/verifier/SoftFingerprint.java
package com.yourco.forge.verifier;

import java.net.NetworkInterface;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Enumeration;
import java.util.HexFormat;
import java.util.Locale;

/**
 * Soft fingerprint — best-effort host identifier used when a license is
 * {@code binding=fingerprint-soft}. Hashes (MAC | hostname | CPU description).
 *
 * <p>Output is a lowercase SHA-256 hex string. The same canonical string format
 * is used across all Forge SDKs (Go / Python / Java / C# / Rust) so a single
 * license can be bound on one host and re-verified by any verifier.
 */
public final class SoftFingerprint {

    private SoftFingerprint() {}

    public static String compute() {
        String mac = safe(primaryMac());
        String hostname = safe(System.getenv().getOrDefault("HOSTNAME",
                System.getenv().getOrDefault("COMPUTERNAME", "")));
        if (hostname.isEmpty()) {
            try {
                hostname = safe(java.net.InetAddress.getLocalHost().getHostName());
            } catch (Exception ignored) {
                hostname = "unknown";
            }
        }
        String cpu = safe(cpuDescription());
        String canonical = (mac + "|" + hostname + "|" + cpu).toLowerCase(Locale.ROOT);
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            return HexFormat.of().formatHex(md.digest(canonical.getBytes(StandardCharsets.UTF_8)));
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException("SHA-256 unavailable on this JVM", e);
        }
    }

    private static String primaryMac() {
        try {
            Enumeration<NetworkInterface> ifaces = NetworkInterface.getNetworkInterfaces();
            while (ifaces.hasMoreElements()) {
                NetworkInterface ni = ifaces.nextElement();
                if (ni.isLoopback() || !ni.isUp() || ni.isVirtual()) continue;
                String name = ni.getName().toLowerCase(Locale.ROOT);
                if (name.startsWith("docker") || name.startsWith("veth")
                        || name.startsWith("vmnet") || name.startsWith("br-")
                        || name.startsWith("utun")) continue;
                byte[] mac = ni.getHardwareAddress();
                if (mac == null || mac.length == 0) continue;
                StringBuilder sb = new StringBuilder(mac.length * 3);
                for (int i = 0; i < mac.length; i++) {
                    if (i > 0) sb.append(':');
                    sb.append(String.format("%02x", mac[i]));
                }
                return sb.toString();
            }
        } catch (Exception ignored) {
            // fall through
        }
        return "";
    }

    private static String cpuDescription() {
        return System.getProperty("os.arch", "") + ":"
                + System.getProperty("os.name", "") + ":"
                + Runtime.getRuntime().availableProcessors();
    }

    private static String safe(String s) {
        return (s == null || s.isEmpty()) ? "unknown" : s;
    }
}
